File: packages/core-ts/src/ws/socket.ts
/**
 * Base WebSocket client over the global `WebSocket` (Node 22+, browsers), or any
 * compatible object handed in through `createWebSocket` (the `ws` package on Node 20, a
 * test double).
 */
import { NetworkError } from '../errors.js'
import { Deferred } from './async.js'

/** What a socket receives: text frames as strings, binary frames as `ArrayBuffer`s. */
export type Data = string | ArrayBuffer

/** The subset of the WebSocket interface this module uses. */
export interface WebSocketLike {
  readonly readyState: number
  binaryType?: string
  send(data: string | ArrayBufferLike | ArrayBufferView): void
  close(code?: number, reason?: string): void
  onopen: ((event: unknown) => void) | null
  onmessage: ((event: { data: unknown }) => void) | null
  onerror: ((event: unknown) => void) | null
  onclose: ((event: { code?: number; reason?: string }) => void) | null
}

export const CONNECTING = 0, OPEN = 1, CLOSING = 2, CLOSED = 3

/** A live connection, plus the promise `wait` races every request against. */
export interface Context {
  readonly ws: WebSocketLike
  /**
   * Never resolves; rejects once the connection is gone -- closed by the peer, failed,
   * a background task threw -- or `abort` was called.
   */
  readonly closed: Promise<never>
  /** End background work and fail `closed` with `reason`; a no-op after the first call. */
  readonly abort: (reason: Error) => void
}

export interface SocketOptions {
  url: string
  /** Milliseconds allowed to open (and to close) the connection. Default 10 s. */
  timeout?: number
  /** Milliseconds between `ping` calls, when `ping` is implemented. Default 24 h. */
  pingInterval?: number
  /** Factory for the raw connection; defaults to the global `WebSocket`. */
  createWebSocket?: (url: string) => WebSocketLike
}

type WebSocketCtor = new (url: string) => WebSocketLike

/**
 * Base WebSocket client: lazy connection, an optional periodic `ping`, a message listener,
 * and error propagation via `wait(...)`.
 *
 * Requires implementing `onMsg`; override `ping` if the server needs application-level
 * pinging.
 *
 * Concurrency contract: the connection opens on first use and closes on `close()` (or
 * `await using`); many concurrent `wait()` calls are fine.
 *
 * The connection can fail without a request noticing: a promise waiting for a reply the
 * socket will never deliver waits forever. `wait(promise)` races it against the
 * connection's own fate, so a dropped connection rejects it with a `NetworkError`.
 */
export abstract class Socket implements AsyncDisposable {
  readonly url: string
  readonly timeout: number
  readonly pingInterval: number
  protected readonly createWebSocket: (url: string) => WebSocketLike
  #opening: Promise<Context> | null = null
  #current: Context | null = null
  #closing: Promise<void> | null = null

  constructor(options: SocketOptions) {
    this.url = options.url
    this.timeout = options.timeout ?? 10_000
    this.pingInterval = options.pingInterval ?? 86_400_000
    this.createWebSocket = options.createWebSocket ?? (url => new ((globalThis as { WebSocket: WebSocketCtor }).WebSocket)(url))
  }

  /** Handle one incoming message. Throwing fails the connection; see `wait`. */
  abstract onMsg(msg: Data): void

  /** Ping the server; called every `pingInterval` while connected, if implemented. */
  ping?(ws: WebSocketLike): Promise<void> | void

  /** Whether a connection is open or being opened. */
  get isOpen(): boolean {
    return this.#opening !== null
  }

  /** The current connection context, opening one first if none exists yet. */
  get ctx(): Promise<Context> {
    return this.open()
  }

  get ws(): Promise<WebSocketLike> {
    return this.open().then(ctx => ctx.ws)
  }

  /** Open the connection, or return the one already open (reconnecting if the peer closed it). */
  async open(): Promise<Context> {
    if (this.#opening) {
      const ctx = await this.#opening
      if (ctx.ws.readyState >= CLOSING) {
        await this.closeContext(ctx)
        return this.open()
      }
      return ctx
    }
    const opening = this.forceOpen().then(
      ctx => { this.#current = ctx; return ctx },
      e => { if (this.#opening === opening) this.#opening = null; throw e },
    )
    this.#opening = opening
    // `return await`, not `return opening`: resolving an async function with a promise
    // costs extra microtask ticks, which would let a concurrent second caller (already
    // awaiting `#opening` above) proceed first and send its frame ahead of this one.
    return await opening
  }

  /**
   * Connect and hand the live connection to the listener and pinger. An override that
   * needs a handshake can `const ctx = await super.forceOpen()` and then `wait(reply, ctx)`.
   */
  protected async forceOpen(): Promise<Context> {
    const ws = await this.connect()
    return this.attach(ws)
  }

  /** Dial `url`; a failure or a `timeout` is a `NetworkError`. */
  protected connect(): Promise<WebSocketLike> {
    return new Promise<WebSocketLike>((resolve, reject) => {
      let ws: WebSocketLike
      try { ws = this.createWebSocket(this.url) } catch (e) {
        return reject(new NetworkError(`Failed to connect to ${this.url}`, { cause: e }))
      }
      const timer = setTimeout(() => { fail(new Error(`timed out after ${this.timeout} ms`)); ws.close() }, this.timeout)
      const fail = (cause: unknown) => {
        clearTimeout(timer)
        reject(new NetworkError(`Failed to connect to ${this.url}`, { cause }))
      }
      ws.onopen = () => { clearTimeout(timer); resolve(ws) }
      ws.onerror = event => fail((event as { error?: unknown })?.error ?? event)
      ws.onclose = event => fail(new Error(`closed (${event.code ?? ''} ${event.reason ?? ''})`.trim()))
      if (ws.readyState === OPEN) { clearTimeout(timer); resolve(ws) }
    })
  }

  /** Wrap an open connection: route frames to `onMsg`, start the pinger, build the `Context`. */
  protected attach(ws: WebSocketLike): Context {
    const closed = new Deferred<never>()
    closed.promise.catch(() => {})
    ws.binaryType = 'arraybuffer'
    const timer = this.ping
      ? setInterval(() => { Promise.resolve().then(() => this.ping!(ws)).catch(abort) }, this.pingInterval)
      : undefined
    ;(timer as { unref?: () => void } | undefined)?.unref?.()
    const abort = (reason: Error) => {
      if (timer !== undefined) clearInterval(timer)
      ws.onmessage = ws.onerror = ws.onclose = null
      closed.reject(reason)
    }
    ws.onmessage = event => {
      try { this.onMsg(event.data as Data) } catch (e) {
        abort(e instanceof Error ? e : new Error(String(e)))
        ws.close()
      }
    }
    ws.onerror = event => abort(new NetworkError('WebSocket error', { cause: (event as { error?: unknown })?.error ?? event }))
    ws.onclose = event => abort(new NetworkError(`Connection closed (${event.code ?? ''} ${event.reason ?? ''})`.trim()))
    return { ws, closed: closed.promise, abort }
  }

  /** Close the connection if one was opened; do nothing when none ever was. */
  async close(): Promise<void> {
    if (!this.#opening) return
    const ctx = await this.#opening.catch(() => null)
    if (ctx) await this.closeContext(ctx)
  }

  async [Symbol.asyncDispose](): Promise<void> {
    await this.close()
  }

  /** Close `ctx`, once, and forget it so the next use opens afresh. */
  protected closeContext(ctx: Context): Promise<void> {
    if (this.#closing) return this.#closing
    if (this.#current === ctx) { this.#current = null; this.#opening = null }
    this.#closing = this.forceClose(ctx).finally(() => { this.#closing = null })
    return this.#closing
  }

  /** Tear down `ctx`: fail every pending `wait`, close the connection and wait for it to end. */
  protected async forceClose(ctx: Context): Promise<void> {
    ctx.abort(new NetworkError('Connection closed'))
    const { ws } = ctx
    if (ws.readyState === CLOSED) return
    await new Promise<void>(resolve => {
      const timer = setTimeout(resolve, this.timeout)
      ;(timer as { unref?: () => void }).unref?.()
      ws.onclose = () => { clearTimeout(timer); resolve() }
      ws.onerror = null
      ws.close()
    })
  }

  /**
   * Wait for `promise`, rejecting instead if the connection fails first.
   *
   * `ctx` defaults to the current connection (opening one if needed); pass it explicitly
   * only from a `forceOpen` override waiting on its own handshake, where the connection
   * being opened is not yet the current one.
   */
  async wait<T>(promise: Promise<T>, ctx?: Context): Promise<T> {
    const live = ctx ?? await this.ctx
    return Promise.race([promise, live.closed])
  }
}

File: packages/core-python/src/truewire_core/ws/socket.py
from typing_extensions import Awaitable, TypeVar
from abc import ABC, abstractmethod
import asyncio
import contextlib
from dataclasses import dataclass, field
from datetime import timedelta
import logging
import websockets

from truewire_core.exceptions import NetworkError

T = TypeVar('T')

logger = logging.getLogger(__name__)

@dataclass
class Context:
  ws: websockets.ClientConnection
  listener: asyncio.Task
  pinger: asyncio.Task

@dataclass
class Socket(ABC):
  """Base WebSocket client.
  
  ### Features
  - Connection handling
  - Optional periodic pinging
  - Message listener
  - Error propagation via `.wait(...)`

  ### Requires implementing
  - `on_msg`: handle incoming messages.
  - `ping`: if you the server requires some custom pinging mechanism

  ### Concurrency Contract
  1. Connection: single owner via `async with`, also supports lazy no-owner use
  2. Requests: many concurrent `wait()` calls OK

  ### Error Propagation

  The websocket connection could fail without you knowing.
  To avoid that happening, you can ensure they are propagated by using `.wait(...)`.

  **Example**:

  ```python
  future = asyncio.Future()

  class MySocket(Socket):
    def on_msg(self, msg: str | bytes):
      future.set_result(msg)

  async with MySocket('wss://example.com') as ws:
    result = await ws.wait(future)
  ```

  If you awaited directly and an exception happened, you would wait forever.
  
  This way, if an error happens, a `NetworkError` will be raised.
  """
  url: str
  timeout: timedelta = field(kw_only=True, default=timedelta(seconds=10))
  ping_interval: timedelta = field(kw_only=True, default=timedelta(hours=24))
  _ctx_future: 'asyncio.Future[Context] | None' = field(default=None, init=False, repr=False)
  """Backing store for `ctx_future`, `None` until something first reaches for it."""
  open_lock: asyncio.Lock = field(default_factory=asyncio.Lock, init=False)
  close_lock: asyncio.Lock = field(default_factory=asyncio.Lock, init=False)

  @property
  def ctx_future(self) -> 'asyncio.Future[Context]':
    """Future holding the connection once one is opened, created on first access.

    Deliberately not a `default_factory`: `asyncio.Future()` binds the running loop as it
    is constructed, so a socket built outside a loop -- a client constructed in a
    synchronous test fixture, or at module scope -- raised `RuntimeError` before anything
    had a chance to connect. A client that owns a socket it may never use has to be
    constructible anywhere; the loop is needed when the socket opens, not when it is built.
    """
    if self._ctx_future is None:
      self._ctx_future = asyncio.Future()
    return self._ctx_future

  @abstractmethod
  def on_msg(self, msg: str | bytes):
    ...

  async def ping(self, ws: websockets.ClientConnection):
    """Ping the server.

    If implemented, this is called periodically by the pinger task.

    Args:
      ws: The live connection, handed directly rather than resolved via `self.ws` —
        `pinger` runs from the moment `force_open` creates it, before `self.ctx` has
        anything to resolve to.
    """
    raise NotImplementedError

  async def pinger(self, ws: websockets.ClientConnection):
    while True:
      await asyncio.sleep(self.ping_interval.total_seconds())
      try:
        await self.ping(ws)
      except NotImplementedError:
        ...

  @property
  async def ctx(self) -> Context:
    """The current connection context, opening one first if none exists yet.

    Reading this property is itself a connect-on-demand: it is right for any caller that
    wants to *use* the socket, but wrong for `__aexit__`, which must be able to close
    without ever opening.
    """
    return await self.open()

  @property
  async def ws(self) -> websockets.ClientConnection:
    return (await self.ctx).ws

  async def __aenter__(self):
    """Take ownership without connecting; the socket opens lazily on first use."""
    return self

  async def __aexit__(self, exc_type, exc_value, traceback):
    """Close the connection if one was opened; do nothing when none ever was.

    Deliberately not `await self.ctx` -- that property falls through to `open()`, so a
    client that only ever used the other transport would dial out here just to hang up.
    """
    if self._ctx_future is not None and self._ctx_future.done():
      await self.close(self._ctx_future.result(), exc_type, exc_value, traceback)
  
  async def force_open(self):
    """Connect, then hand the live connection straight to `listener`/`pinger`.

    Deliberately not `self.listener()`/`self.pinger()` resolving `ws` themselves via
    `self.ws` — that property goes through `self.ctx` -> `self.open()`, which is *this
    call*, still in progress. A subclass override that needs to send-and-await a reply
    during its own `force_open` (an auth handshake, say) would otherwise deadlock: the
    listener can't start delivering until `self.ctx_future` resolves, and it only
    resolves once `force_open` returns — which it can't, waiting on a reply only the
    listener can deliver. Passing `ws` as a parameter here means `listener`/`pinger`
    never need `self.ctx` at all, so there's nothing left for them to wait on.
    """
    async def connect():
      try:
        return await websockets.connect(self.url, open_timeout=self.timeout.total_seconds())
      except websockets.exceptions.WebSocketException as e:
        raise NetworkError(f'Failed to connect to {self.url}') from e

    ws = await connect()
    logger.info('Connected!')
    return Context(
      ws=ws,
      listener=asyncio.create_task(self.listener(ws)),
      pinger=asyncio.create_task(self.pinger(ws)),
    )
  
  async def open(self):
    if self.open_lock.locked() or self.ctx_future.done():
      ctx = await self.ctx_future
      if ctx.ws.state == websockets.State.CLOSED:
        await self.close(ctx)
        return await self.open()
      else:
        return ctx

    async with self.open_lock:
      logger.info('Connecting...')
      ctx = await self.force_open()
      self.ctx_future.set_result(ctx)
      return ctx

  async def force_close(self, ctx: Context, exc_type=None, exc_value=None, traceback=None):
    ctx.listener.cancel()
    ctx.pinger.cancel()
    await ctx.ws.__aexit__(exc_type, exc_value, traceback)

  async def close(self, ctx: Context, exc_type=None, exc_value=None, traceback=None):
    if not self.close_lock.locked():
      async with self.close_lock:
        self._ctx_future = None
        await self.force_close(ctx, exc_type, exc_value, traceback)

  async def listener(self, ws: websockets.ClientConnection):
    while True:
      try:
        msg = await ws.recv()
        logger.debug('Received: %s', msg)
        self.on_msg(msg)
      except websockets.exceptions.WebSocketException as e:
        logger.error('Error receiving message: %s', e)
        raise NetworkError('Error receiving message') from e

  async def wait(self, fut: Awaitable[T], *, ctx: Context | None = None) -> T:
    """Wait for a future to complete, propagating any exceptions in the background tasks.

    Args:
      fut: Future to wait for.
      ctx: Connection context to race against, if already held. Defaults to resolving
        `self.ctx` -- the only exception is a `force_open` override waiting on its own
        bootstrap reply, where `self.ctx` isn't resolved yet (it resolves through this
        same `force_open` call, still in progress) but the override already has the
        `Context` it just built, from `ctx = await super().force_open()`.
    """
    if ctx is None:
      ctx = await self.ctx
    async def coro():
      return await fut
    task = asyncio.create_task(coro())
    done, _ = await asyncio.wait([task, ctx.listener, ctx.pinger], return_when='FIRST_COMPLETED')
    if task not in done:
      # The listener/pinger won the race instead -- `task` is still pending
      # and nothing else will ever await or cancel it, so it would otherwise
      # be silently garbage-collected mid-flight ("Task was destroyed but it
      # is pending!").
      task.cancel()
      with contextlib.suppress(asyncio.CancelledError):
        await task
    if ctx.listener in done:
      if (exc := ctx.listener.exception()) is not None:
        raise exc
      else:
        raise RuntimeError('Listener task ended unexpectedly')
    elif ctx.pinger in done:
      if (exc := ctx.pinger.exception()) is not None:
        raise exc
      else:
        raise RuntimeError('Pinger task ended unexpectedly')
    return task.result()
